crates/red_knot_python_semantic/src/types.rs

```diff
@@ -29,6 +29,77 @@ mod display;
 mod infer;
 mod narrow;
 
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub enum Boundness {
+    Bound,
+    MayBeUnbound,
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub enum Symbol<'db> {
+    Type(Type<'db>, Boundness),
+    Unbound,
+}
+
+impl<'db> Symbol<'db> {
+    pub fn is_unbound(&self) -> bool {
+        matches!(self, Symbol::Unbound)
+    }
+
+    #[must_use]
+    pub fn replace_unbound_with(self, db: &'db dyn Db, replacement: &Symbol<'db>) -> Symbol<'db> {
+        match replacement {
+            Symbol::Type(replacement, _) => Symbol::Type(
+                match self {
+                    Symbol::Type(ty, Boundness::Bound) => ty,
+                    Symbol::Type(ty, Boundness::MayBeUnbound) => {
+                        UnionType::from_elements(db, [*replacement, ty])
+                    }
+                    Symbol::Unbound => *replacement,
+                },
+                Boundness::Bound,
+            ),
+            Symbol::Unbound => self,
+        }
+    }
+
+    #[cfg(test)]
+    #[track_caller]
+    fn expect_bound(self) -> Type<'db> {
+        match self {
+            Symbol::Type(ty, Boundness::Bound) => ty,
+            _ => {
+                panic!("Expected a bound type")
+            }
+        }
+    }
+
+    fn unwrap_or(&self, other: Type<'db>) -> Type<'db> {
+        match self {
+            Symbol::Type(ty, _) => *ty,
+            Symbol::Unbound => other,
+        }
+    }
+
+    fn as_type(&self) -> Option<Type<'db>> {
+        match self {
+            Symbol::Type(ty, _) => Some(*ty),
+            Symbol::Unbound => None,
+        }
+    }
```

Comment:
I know that @MichaReiser asked for this method and @carljm liked Micha's comment... I worry that this blurs the line between bound symbols and possibly-but-not-necessarily bound symbols. In some ways, possibly-unbound symbols are more similar to unbound symbols than they are similar to always-bound symbols, in that attempts to read them should _always_ cause us to emit a diagnostic. The fact that we currently have a blurred line betwee possibly-bound and always-bound symbols is exactly what causes the bugs outlined in https://github.com/astral-sh/ruff/issues/14012.

If there are situations where we really need to ignore that a symbol might possibly be unbound, I think I'd honestly prefer to be explicit about that by manually matching on the variants, rather than using this method that collapses it into a binary "either you have a `Type`, or you don't" enum.

Reply:
(I agree with what you said in standup earlier about merging this PR sooner rather than later to avoid merge conflicts. Don't let this comment block that from happening! This can always be tackled as a followup.)

Reply:
I actually had the same thought. And considered calling this function something like `Symbol::ignore_possibly_unbound`. I'll think about it and probably do a follow-up.

Reply:
I mentioned this in another comment, but this issue isn't specific to `as_type`. Basically all of the methods on Symbol currently ignore possibly-unboundness in some form.

Reply:
> in that attempts to read them should _always_ cause us to emit a diagnostic

I don't think this is always true. For example, if `__iadd__` is possibly unbound we should union its return type with the return type from `__add__`, and only emit a diagnostic if `__add__` doesn't exist.

But I do agree that callers should be the ones make the decision about suppressing such diagnostics, and the APIs should be very clear if they do that.

Reply:
> I don't think this is always true. For example, if `__iadd__` is possibly unbound we should union its return type with the return type from `__add__`, and only emit a diagnostic if `__add__` doesn't exist.

Right, that's a useful correction. But even here, we still need to account for the possibly-unbound nature and treat it differently to the case where it was definitely bound.